src/eslint_result.ts

```diff
@@ -0,0 +1,98 @@
+import * as core from "@actions/core"
+import { ChangeRange, generateChangeRanges } from "./git_utils"
+
+export type ResultObject = {
+  filePath: string
+  messages: EslintMessage[]
+}
+
+export type EslintResultInstance = InstanceType<typeof EslintResult>
+
+type EslintMessage = {
+  ruleId: string
+  severity: number
+  message: string
+  line: number
+  column: number
+  nodeType: string
+  messageId: string
+  endLine: number
+  endColumn: number
+  suggestions: any[]
+}
+
+export class EslintResult {
+  public relevantWarningCount: number = 0
+  public relevantErrorCount: number = 0
+  private resultObject: ResultObject
+  private compareSha: string
+  private changeRanges?: ChangeRange[]
+  private relevantMessages: EslintMessage[] = []
+
+  constructor(resultObject: ResultObject, compareSha: string) {
+    this.resultObject = resultObject
+    this.compareSha = compareSha
+  }
+
+  async asyncInitialize() {
+    this.changeRanges = await generateChangeRanges(
+      this.resultObject.filePath,
+      this.compareSha,
+    )
+    this.findRelevantMessages()
+    this.calculateCounts()
+  }
+
+  outputAnnotations() {
+    this.relevantMessages.forEach((msg) => {
+      let options: core.AnnotationProperties = {
+        title: msg.ruleId,
+        file: this.repoFilePath,
+        startLine: msg.line,
+        endLine: msg.endLine,
+        startColumn: msg.column,
+        endColumn: msg.endColumn,
+      }
+      switch (msg.severity) {
+        case 1:
+          core.warning(msg.message, options)
+          break
+        case 2:
+          core.error(msg.message, options)
+        default:
+          break
+      }
+    })
+  }
+
+  private findRelevantMessages() {
+    this.relevantMessages = this.messages.filter((m) =>
+      this.changeRanges?.some((changeRange) => changeRange.doesInclude(m.line)),
+    )
+  }
+
+  private calculateCounts() {
+    this.relevantMessages.forEach((msg) => {
+      switch (msg.severity) {
+        case 1:
+          this.relevantWarningCount += 1
```

Comment:
The imperative nature of this class has me wondering if we could find a more imperative mental model to generate the results we want. I wonder what it would look like if we waited to initialize the `EslintResult` until we'd generated the data we care about. Maybe we could have a `static async` function on `EslintResult` that transforms the `ResultObject` into a fully-formed `EslintResult`.

It might be something like

```ts
class EslintResult {
    static async function fromResult(result: ResultObject, compareSha: string) {
        const changeRanges = await generateChangeRanges(result.filePath, compareSha)
        const relevantMessages = result.messages.filter(m => changeRanges.some(cr => cr.doesInclude(m.line)))
        const { warningCount, errorCount } = calculateCounts(relevantMessages)

        return new EslintResult({ result, relevantMessages, warningCount, errorCount })
    }

    static function calculateCounts(messages: ResultObject['messages']) {
        return messages.reduce((acc, message) => {
            switch (message.severity) {
                case 1:
                    acc.warningCount += 1
                    break
                case 2:
                    acc.errorCount += 1
                    break
            }

            return acc
        }, { warningCount: 0, errorCount: 0 })
    }
}
```

We can still initialize all the objects async (with `Promise.all`), but we no longer need to call instance methods in a certain order for them to be valid.

Reply:
I like that idea, I'll take a look 👍